app/models/search_facet.rb

```diff
@@ -0,0 +1,165 @@
+class SearchFacet
+  include Mongoid::Document
+  include Mongoid::Timestamps
+
+  extend ErrorTracker
+
+  field :name, type: String
+  field :identifier, type: String
+  field :filters, type: Array, default: []
+  field :is_ontology_based, type: Boolean, default: false
+  field :ontology_urls, type: Array, default: []
+  field :is_array_based, type: Boolean
+  field :big_query_id_column, type: String
+  field :big_query_name_column, type: String
+  field :convention_name, type: String
+  field :convention_version, type: String
+
+  validates_presence_of :name, :identifier, :big_query_id_column, :big_query_name_column, :convention_name, :convention_version
+  validates_uniqueness_of :big_query_id_column, scope: [:convention_name, :convention_version]
+  validate :ensure_ontology_url_format, if: proc {|attributes| attributes[:is_ontology_based]}
+  before_create :set_is_array_based, if: proc {|attributes| attributes[:is_array_based].nil?}
+
+  def self.big_query_dataset
+    ApplicationController.big_query_client.dataset(CellMetadatum::BIGQUERY_DATASET)
+  end
+
+  # retrieve table schema definition
+  def self.get_table_schema(table_name: CellMetadatum::BIGQUERY_TABLE, column_name: nil)
+    begin
+      query_string = "SELECT column_name, data_type, is_nullable FROM INFORMATION_SCHEMA.COLUMNS WHERE table_name='#{table_name}'"
+      schema = self.big_query_dataset.query(query_string)
+      if column_name.present?
+        schema.detect {|column| column[:column_name] == column_name}
+      else
+        schema
+      end
+    rescue => e
+      Rails.logger.error "Error retrieving table schema for #{CellMetadatum::BIGQUERY_TABLE}: #{e.class.name}:#{e.message}"
+      error_context = ErrorTracker.format_extra_context({query_string: query_string})
+      ErrorTracker.report_exception(e, nil, error_context)
+      []
+    end
+  end
+
+  # update all search facet filters after BQ update
+  def self.update_all_facet_filters
+    self.all.each do |facet|
+      Rails.logger.info "Updating #{facet.name} filter values"
+      updated = facet.update_filter_values!
+      if updated
+        Rails.logger.info "Update to #{facet.name} complete!"
+      else
+        Rails.logger.error "Update to #{facet.name} failed"
+      end
+    end
+  end
+
+  # return a formatted object to use to render search UI component
```

Comment:
Nice, we'll just want to move this method to the controller once we make the API